File: asherah-dotnet-shared/src/Crypto/CryptoPolicy.cs
using System;
using GoDaddy.Asherah.Crypto.ExtensionMethods;

namespace GoDaddy.Asherah.Crypto;

public abstract class CryptoPolicy
{
    public enum KeyRotationStrategy
    {
        Inline,
        Queued,
    }

    public abstract bool IsKeyExpired(DateTimeOffset keyCreationDate);
    public abstract long GetRevokeCheckPeriodMillis();
    public abstract bool CanCacheSystemKeys();
    public abstract bool CanCacheIntermediateKeys();
    public abstract bool CanCacheSessions();
    public abstract long GetSessionCacheMaxSize();
    public abstract long GetSessionCacheExpireMillis();
    public abstract bool NotifyExpiredIntermediateKeyOnRead();
    public abstract bool NotifyExpiredSystemKeyOnRead();
    public abstract KeyRotationStrategy GetKeyRotationStrategy();

    public virtual bool IsInlineKeyRotation() => GetKeyRotationStrategy() == KeyRotationStrategy.Inline;
    public virtual bool IsQueuedKeyRotation() => GetKeyRotationStrategy() == KeyRotationStrategy.Queued;

    public virtual DateTimeOffset TruncateToSystemKeyPrecision(DateTimeOffset dateTimeOffset) =>
        dateTimeOffset.Truncate(TimeSpan.FromMinutes(1));

    public virtual DateTimeOffset TruncateToIntermediateKeyPrecision(DateTimeOffset dateTimeOffset) =>
        dateTimeOffset.Truncate(TimeSpan.FromMinutes(1));
}

File: asherah-java/java/src/main/java/com/godaddy/asherah/crypto/CryptoPolicy.java
package com.godaddy.asherah.crypto;

import com.godaddy.asherah.jni.AsherahConfig;

/**
 * Abstract crypto policy. Compatible with the canonical godaddy/asherah CryptoPolicy.
 * In the FFI binding, policy settings are mapped to AsherahConfig fields.
 */
public abstract class CryptoPolicy {

    public enum KeyRotationStrategy {
        INLINE,
        QUEUED
    }

    public abstract boolean isKeyExpired(java.time.Instant keyCreationDate);
    public abstract long getRevokeCheckPeriodMillis();
    public abstract boolean canCacheSystemKeys();
    public abstract boolean canCacheIntermediateKeys();
    public abstract boolean canCacheSessions();
    public abstract long getSessionCacheMaxSize();
    public abstract long getSessionCacheExpireMillis();
    public abstract boolean notifyExpiredIntermediateKeyOnRead();
    public abstract boolean notifyExpiredSystemKeyOnRead();
    public abstract KeyRotationStrategy keyRotationStrategy();

    public boolean isInlineKeyRotation() {
        return keyRotationStrategy() == KeyRotationStrategy.INLINE;
    }

    public boolean isQueuedKeyRotation() {
        return keyRotationStrategy() == KeyRotationStrategy.QUEUED;
    }

    /**
     * Apply this policy's settings to an AsherahConfig builder.
     * Package-private — called by SessionFactory during build().
     */
    public void applyConfig(final AsherahConfig.Builder builder) {
        if (canCacheSessions()) {
            builder.enableSessionCaching(true);
            builder.sessionCacheMaxSize((int) getSessionCacheMaxSize());
            builder.sessionCacheDuration(getSessionCacheExpireMillis() / 1000);
        } else {
            builder.enableSessionCaching(false);
        }
    }
}
